Basic SSA Validation

umar456 opened the pull request:
* Updated validation tests to assignment form
* Fixed function tests to handle multiple parameters
* Basic SSA Validation

AWoloszyn commented on source/validate.cpp:
```diff
@@ -166,14 +171,15 @@ spv_result_t spvValidateBasic(const spv_instruction_t* pInsts,
       } else if (SPV_OPERAND_TYPE_LITERAL_INTEGER == type) {
-        spvCheckReturn(spvValidateOperandsLiteral(
-            words + index, wordCount - index, 2, position, pDiagnostic));
+        // spvCheckReturn(spvValidateOperandsNumber(
```
If you plan on re-enabling this code, then use:
#if 0
#endif
Around disabled code, and add a TODO(): to re-enable at some point

Otherwise just delete the code.

umar456 replied:
Fixed da35d28

dneto0 commented on source/validate.cpp:
```diff
@@ -31,6 +31,10 @@
 #include <string.h>
 
 #include <vector>
+#include <iterator>
```
We sort these. Not a big deal, we can come back to this.  The rest of the code is not fully compliant either.
For reference: https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

antiagainst replied:
Now `clang-format` can do this automatically for you, I think.

umar456 replied:
Fixed da35d28

dneto0 commented on source/validate.cpp:
```diff
@@ -31,6 +31,10 @@
+#include <iterator>
+#include <set>
+
+using std::vector;
```
The using declarations should come after all #includes.  Otherwise you might mess up the type mappings seen by those other includes.

umar456 replied:
Fixed da35d28

dneto0 commented on source/validate.cpp:
```diff
@@ -80,6 +84,7 @@ spv_result_t spvValidateOperandValue(const spv_operand_type_t type,
                                      spv_diagnostic* pDiagnostic) {
   switch (type) {
     case SPV_OPERAND_TYPE_ID:
+    case SPV_OPERAND_TYPE_TYPE_ID:
```
There's also: SPV_OPERAND_TYPE_MEMORY_SEMANTICS_ID, SPV_OPERAND_TYPE_SCOPE_ID.

Since you started working on this, I've reorganized the enumerants in spv_operand_type_t so these groupings are easier to see.

antiagainst commented on source/validate.cpp:
```diff
@@ -261,6 +267,143 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+    return err;
+  }
+
+  return SPV_SUCCESS;
```
I believe L398 - L404 can be combined into one instruction `return SSAPass(...)`?

antiagainst commented on source/validate_id.cpp:
```diff
@@ -1216,12 +1216,11 @@ bool idUsage::isValid<SpvOpFunctionParameter>(const spv_instruction_t* inst,
-    spvCheck(SpvOpFunction != function->opcode &&
-                 SpvOpFunctionParameter != function->opcode,
+  while (firstInst != --function) {
+    spvCheck(SpvOpFunction != function->opcode && SpvOpFunctionParameter != function->opcode,
```
If possible, I'd suggest to remove `spvCheck`. It's just an if-statement defined as a `macro`...

antiagainst commented on source/validate.cpp:
```diff
@@ -261,6 +267,143 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+  validation_state_t(const uint32_t* binary, spv_diagnostic* diag)
+      : binary(binary), diagnostic(diag), program_counter(binary) {}
+
+  void set_instruction_words(const uint32_t * const words) {
```
`*` alignment. Make sure to run `clang-format` on the file. :)

dneto0 replied:
Much as I dislike camelcase (and I really do), convention is to use camelcase for method names.
Style guide says method names start with a captial.  The rest of the SPIR-V Tools hasn't caught up to that yet.

antiagainst commented on source/validate.cpp:
```diff
@@ -261,6 +267,143 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+    return SPV_SUCCESS;
+  }
+
+  size_t forward_declared_id_count() { return forward_ids.size(); }
```
`const` function.

dneto0 replied:
unresolved_forward_id_count(), or num_unresolved_forward_ids()

antiagainst commented on source/validate.cpp:
```diff
@@ -261,6 +267,143 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+
+  size_t forward_declared_id_count() { return forward_ids.size(); }
+
+  bool check_declared_id(uint32_t id) {
```
`const` function.

dneto0 replied:
"check" is ambiguous.  Prefer "isDefinedId"

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +267,143 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+  spv_diagnostic* diagnostic;
+  const uint32_t* program_counter;
+
+  std::set<uint32_t> declared_ids;
```
Prefer std::unordered_set.  Regular sets are usually implemented as trees of nodes which use up a lot of space and are slow since there's so much pointer chasing.
(Just use the default hash function.)

dneto0 replied:
Also, these are class data members, so they should go at the end.

umar456 replied:
Fixed da35d28

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +267,143 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+}
+
+// TODO(umar): Move this class to another file
+class validation_state_t {
```
Class names in CamelCase

umar456 replied:
Fixed da35d28

antiagainst commented on source/validate.cpp:
```diff
@@ -261,6 +267,143 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+  const uint32_t* inst_ptr = _.get_instruction_ptr();
+
+  for (int i = 0; i < inst->num_operands; i++) {
+    spv_parsed_operand_t operand = inst->operands[i];
```
`const` variables.

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +267,143 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+class validation_state_t {
+  const uint32_t* binary;
+  spv_diagnostic* diagnostic;
+  const uint32_t* program_counter;
```
What is the program counter?  What's it counting?  I presume it's referencing someone else's storage?

dneto0 replied:
From usage, seems this should be instruction_words_
?

umar456 replied:
It was a reference to the [program counter](https://en.wikipedia.org/wiki/Program_counter) register in a CPU. Its probably confusing in this context. I can change it to `instruction_ptr_` or `instruction_word_`

antiagainst commented on source/validate.cpp:
```diff
@@ -261,6 +267,143 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+
+  const uint32_t* inst_ptr = _.get_instruction_ptr();
+
+  vector<SpvOp_> instructions_with_forward_ids = {
```
`SpvOp` instead of `SpvOp_`?

dneto0 replied:
It's cleaner and likely faster to use a separate method  bool InstructionAllowsForwardIdReferences(SpvOp) that just has a switch statement with the positive cases, a default that breaks out, and a default return of false.

umar456 replied:
This was a placeholder. I have updated this to return a functor based off of the opcode and return true or false based on the index of the operand. See 3949bf4

https://github.com/KhronosGroup/SPIRV-Tools/commit/3949bf47a732c935d49a5e0886a886a7a1d7f1b7#diff-b8a71d60452cadda1e5f190db6789d2dR411

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +267,143 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+    return declared_ids.find(id) != std::end(declared_ids);
+  }
+
+  const uint32_t* get_binary() const { return binary; }
```
Not used?  Remove it, I think.

umar456 replied:
Fixed da35d28

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +267,143 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+
+// TODO(umar): Move this class to another file
+class validation_state_t {
+  const uint32_t* binary;
```
Names of class data members should end in "_".
Declarations in a struct/class should be public:, then protected:, then private:.
So these go at the end and are binary_, diagnostic_, and program_counter_.

dneto0 replied:
I think the "binary" member is only used by the binary() method, which itself is never used.  Remove it, I think.

umar456 replied:
This was a relic which was based on an older version of spy_parsed_instruction_t struct which did not have the words pointer.

Fixed da35d28

antiagainst commented on source/validate.cpp:
```diff
@@ -261,6 +267,143 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+  vector<SpvOp_> instructions_with_forward_ids = {
+      SpvOpExecutionMode, SpvOpEntryPoint, SpvOpPhi, SpvOpFunctionCall,
+      SpvOpName, SpvOpMemberName, SpvOpDecorate, SpvOpMemberDecorate,
+      SpvOpGroupDecorate};
```
Missing `SpvOpGroupMemberDecorate`?

umar456 replied:
Fixed 3949bf4

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +267,143 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+  }
+
+  spv_result_t declare_id(uint32_t id) {
+    if (declared_ids.find(id) == std::end(declared_ids)) {
```
I'd usually write declared_ids.end().  Seems more direct. ?

umar456 replied:
Following http://herbsutter.com/elements-of-modern-c-style/

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +267,143 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+    program_counter = words;
+  }
+
+  spv_result_t declare_id(uint32_t id) {
```
All methods should have a description before their declaration. E.g. for this one 
// Records the given Id as having been declared.

dneto0 replied:
Actually, aren't these definitions, not declarations?

umar456 replied:
Good point. Fixed da35d28

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +267,143 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
   return SPV_SUCCESS;
 }
 
+// TODO(umar): Validate header
```
A subtle thing:  the Id bound should be validated also.  But you can only do that after you've seen all the instructions in the module.

dneto0 replied:
FYI. The binary parser validates the magic word, and the length of the header, but nothing else.

umar456 replied:
Added notes da35d28

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +267,143 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+  const uint32_t* program_counter;
+
+  std::set<uint32_t> declared_ids;
+  std::set<uint32_t> forward_ids;
```
From usage, this isn't really forward_ids.  It's unresolved_forward_ids.  That is, it's the set of Ids for which we've seen a forward declaration but no definition.  I think the name needs to be fixed, and we need a comment to describe it.

umar456 replied:
Good point. Fixed da35d28

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +267,143 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+  validation_state_t& _ = *(reinterpret_cast<validation_state_t*>(user_data));
+  _.set_instruction_words(inst->words);
+
+  const uint32_t* inst_ptr = _.get_instruction_ptr();
```
Here, inst_ptr is just inst->words.  No need to go through a method to get it out again.

umar456 replied:
Fixed da35d28

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +267,143 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+  const uint32_t* inst_ptr = _.get_instruction_ptr();
+
+  vector<SpvOp_> instructions_with_forward_ids = {
+      SpvOpExecutionMode, SpvOpEntryPoint, SpvOpPhi, SpvOpFunctionCall,
```
Also:
  OpBranch, OpBranchConditional,  (See section 2.5)
  OpTypeForwardPointer
  Any instruction with an "Invoke" parameter, since the Invoke parameter is an OpFunction.  As 2.4 says "This allows for recursion and early declaration of entry points."  So this list includes: OpEnqueueKernel, OpGetKernelNDRangeSubGroupCount, OpGetKernelNDRangeMaxSubGroupSize, OpGetKernelWorkGroupSize, OpGetKernelPreferredWorkGroupSizeMultiple

dneto0 replied:
OpSwitch is another branching instruction.  But its selector Id must be defined (can't be a forward reference).  Its default Id and case label Ids can be forward references.

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +267,143 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+        ret = _.declare_id(*operand_ptr);
+        break;
+      }
+      case SPV_OPERAND_TYPE_TYPE_ID:
```
There are two other ID kinds you have to check: SPV_OPERAND_TYPE_MEMORY_SEMANTICS_ID, and SPV_OPERAND_TYPE_SCOPE_ID.

umar456 replied:
Fixed da35d28

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +267,143 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+      }
+      case SPV_OPERAND_TYPE_TYPE_ID:
+      case SPV_OPERAND_TYPE_ID: {
+        if (can_have_forward_declared_ids) {
```
For OpBranchConditional, this is not quite right.  The condition Id must have been defined by now.  It's the true branch and false branch that could be forward references.

The general point is that whether an Id can be forward declared is a function of both opcode and argument index.

umar456 replied:
Fixed in 3949bf4

dneto0 commented on source/validate_id.cpp:
```diff
@@ -1216,12 +1216,11 @@ bool idUsage::isValid<SpvOpFunctionParameter>(const spv_instruction_t* inst,
-  while (firstInst != function) {
-    spvCheck(SpvOpFunction != function->opcode &&
-                 SpvOpFunctionParameter != function->opcode,
+  while (firstInst != --function) {
```
This is dangerous.  A malicious input will cause this code to read out of bounds.

umar456 replied:
Fixed using assertion in da35d28

dneto0 commented on source/validate_id.cpp:
```diff
@@ -1216,12 +1216,11 @@ bool idUsage::isValid<SpvOpFunctionParameter>(const spv_instruction_t* inst,
                                    << inst->words[resultTypeIndex]
                                    << "' is not defined.";
```
FYI. The spvValidateIDs function in validate.cpp needs to be reimplemented around the spvBinaryParse and spv_parsed_instruction_t because the technique of using spvBinaryOperandInfo function is not expressive enough to describe all the types of instructions, and the fact that they can be variable length.  For example OpSwitch has variable number of operands, and some of them are literal numbers, and others are label Ids. 
I think the entire validate_id.cpp file can be shunk to about 100 lines of code by using a data-driven approach around the spv_parsed_instruction_t.

umar456 replied:
That was my thought. I didn't want to break all of the ValidateID unit tests in this commit. That is something I have been prototyping in another branch.

dneto0 replied:
+1

dneto0 commented on source/validate_id.cpp:
```diff
@@ -1216,12 +1216,11 @@ bool idUsage::isValid<SpvOpFunctionParameter>(const spv_instruction_t* inst,
                                    << "' is not defined.";
            return false);
-  auto function = inst - 1;
+  auto function = inst;
```
This seems kind of pointless.  Just go ahead and rename the parameter inst to function if you like.

dneto0 commented:
Some bugs, missing cases.
I didn't look too closely at the tests, but they look inadequate for the scenarios you're trying to cover in this patch.

Also, the perennial style issues. :-)

umar456 commented:
I agree. Tests are slowly coming along because I am trying to be careful to create correct SPIR-V code. Its hard to find reference code for some of the instructions :) I should be faster once I am more familiar with the language.

dneto0 commented on CMakeLists.txt:
```diff
@@ -79,7 +79,6 @@ function(default_compile_options TARGET)
       target_compile_options(${TARGET} PRIVATE -fno-omit-frame-pointer)
     endif()
     if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
-      target_compile_options(${TARGET} PRIVATE -fcolor-diagnostics)
```
Ideally this would have been put under a configuration option.  But meh, I don't care enough.

umar456 replied:
I mentioned this in the commit message. This is clang's default argument for terminals that support color. Terminals which do not support color display ASCII escape codes if that option is included. This was causing issues in Emacs which automatically converts the line output into links so color is not necessary.

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +273,190 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+  int incrementInstructionCount() { return instruction_counter_++; }
+
+  libspirv::DiagnosticStream diag(spv_result_t error_code) const {
+    return libspirv::DiagnosticStream({0, 0, (uint64_t)(instruction_counter_)},
```
The convention in the binary parser / disassembler is that the "index" field of the diagnostic is the word number, not the instruction number.    But unfortunately the spv_position_t doesn't make it clear what it should be.  :-(
I think word offset is more helpful because it tells you were to look in a binary dump.  But I'll open an issue to resolve the question.

dneto0 replied:
Created https://github.com/KhronosGroup/SPIRV-Tools/issues/31

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +273,190 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+  unordered_set<uint32_t> unresolved_forward_ids_;
+};
+
+spv_result_t SSAPass(ValidationState_t& _,
```
You really need a description for this, especially because can_have_forward_declared_ids is non-trivial.

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +273,190 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+  return SPV_SUCCESS;
+}
+
+template <SpvOp OP>
```
Needs description.  This one and the other similar ones.
The other ones will need to reference the grammar for the operand numbers.

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +273,190 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+
+template <>
+bool can_forward_declare<SpvOpPhi>(int index) {
+  return (1 + index % 2);
```
??

Should be:   return (i>=2)

You'll need some comments to say why.

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +273,190 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+};
+
+spv_result_t SSAPass(ValidationState_t& _,
+                     function<bool(int)> can_have_forward_declared_ids,
```
Since the argument index is never negative, I'd make it function<bool(unsigned)>

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +273,190 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+spv_result_t SSAPass(ValidationState_t& _,
+                     function<bool(int)> can_have_forward_declared_ids,
+                     const spv_parsed_instruction_t* inst) {
+  for (int i = 0; i < inst->num_operands; i++) {
```
and by extension, i should probably be unsigned.

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +273,190 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+  switch (opcode) {
+    case SpvOpExecutionMode:
+    case SpvOpEntryPoint:
+    case SpvOpFunctionCall:
```
only the called function can be forward declared.  not the result id nor the arguments.

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +273,190 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+    case SpvOpFunctionCall:
+    case SpvOpName:
+    case SpvOpMemberName:
+    case SpvOpSelectionMerge:
```
OpLoopMerge

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +273,190 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+    case SpvOpDecorate:
+    case SpvOpMemberDecorate:
+    case SpvOpBranch:
+      out = can_forward_declare<SpvOpNop>;
```
The use of SpvOpNop is confusing here because OpNop isn't captured by any of the cases that execute this code.  May as well just use the plain lambda inline:   out = [](int){ return true; }

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +273,190 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+      break;
+
+    default:
+      out = [](int index) {
```
Just don't name the argument and you don't have to write " (void)index;'

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +273,190 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+  return (1 + index % 2);
+}
+
+template <>
```
Actually these templated functions aren't buying you much.
For instance, the OpBranchConditional and OpGroupDecorate functionality are the same.  Yet you reuse the same function for OpGroupDecorate and OpDecorate.

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +273,190 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+      out = can_forward_declare<SpvOpGroupDecorate>;
+      break;
+
+    case SpvOpBranchConditional:
```
Can collapse this case into the same sequence as OpGroupDecorate.
And you could avoid making the templated function and instead write:
   out = [](int index){ return index != 0; }

Your choice whether to keep the lambdas, but consolidate cases where you can. (And document them.)

dneto0 replied:
Missing OpSwitch. It's same as OpBranchConditional.

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +273,190 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+
+    auto ret = SPV_ERROR_INTERNAL;
+    switch (type) {
+      case SPV_OPERAND_TYPE_RESULT_ID: {
```
No need for the braces since you're not defining or declaring variables.  But you can keep if you like.

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +273,190 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+  return out;
+}
+
+spv_result_t pushInstructions(void* user_data,
```
Needs a description.
In what sense is this a "push"?  The best I can think of is pushing an instruction through a pipeline of passes.
But you haven't described that.  Either say that's your intent, and that further passes will be added in this function.
Or just call it ProcessInstruction.

dneto0 commented on source/validate.cpp:
```diff
@@ -279,6 +475,19 @@ spv_result_t spvValidate(const spv_const_context context,
     return SPV_ERROR_INVALID_BINARY;
   }
 
+  ValidationState_t vstate(pDiagnostic);
```
It's not clear what the evolution of the algorithm is going to be.
I suspect you're going to do per-instruction processing in a single spvBinaryParse, but also collect info to be checked after the fact.  Should give that high level view here.

dneto0 commented on source/validate.cpp:
```diff
@@ -279,6 +475,19 @@ spv_result_t spvValidate(const spv_const_context context,
+  auto err = spvBinaryParse(context, &vstate, binary->code, binary->wordCount,
+                            setHeader, pushInstructions, pDiagnostic);
+
+  if (err) {
```
And here say "Now validate things which require the entire module to have been seen.  Use the data we collected on the single parsing pass through the module."
But only if that's what the design is going to be.  :-)

dneto0 commented on source/validate_id.cpp:
```diff
@@ -1216,22 +1216,21 @@ bool idUsage::isValid<SpvOpFunctionParameter>(const spv_instruction_t* inst,
-    spvCheck(SpvOpFunction != function->opcode &&
-                 SpvOpFunctionParameter != function->opcode,
+  assert(firstInst < inst && "Invalid instruction pointer");
+  while (firstInst != --inst) {
```
Ok.  I see that firstInst is initialized to the start of the copied instructions array.  So this firstInst != --inst actually will terminate, and won't result in an out-of-bounds access.

dneto0 commented:
Heading in the right direction.
Some issues with the the by-operand checks for forward ID references.

For the validation tests, I encourage you to make a regular structure so that it will be easy to see (or predict) where a test will go, and see if one is missing.  Use multiple test files to organize them. E.g. Validate.SSA.cpp for forward ID checks and general SSA checks.

umar456 commented:
Addressed the feedback so far

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +273,173 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+// Performs SSA validation on the IDs of an instruction. The
+// can_have_forward_declared_ids  functor should return true if the
+// instruction operand's ID can be forward referenced.
+spv_result_t SsaPass(ValidationState_t& _,
```
I just realized this strategy won't work.
The problem is it never computes dominators in the control flow.
The trouble is that your SsaPass algorithm will think any result-id for a previous instruction is valid to use.  But it doesn't distinguish between previous instructions that dominate this use vs. those that don't.

Here's an example where it would fail.  It's roughly structured as an if-then-else block, but where a value created in the "then" block is used in the "else" block.


```
   %conditionBlock = Label 
                                  SelectionMerge %merge None
                                  BranchConditional %condition %thenBlock %elseBlock

  %thenBlock = Label
                         %CreatedInThen = OpIAdd %i32 %one %two
                         Branch %merge

 %elseBlock  = Label
                        %BAD = OpCopyObject %CreatedInThen         ;;;;;; This is wrong. 
                                                                ;;;;;;;  Since %CreatedInThen definition does not dominate this basic block.
                        Branch %merge

   %merge = Label
  
```

Even more severe, a value defined in one function body would be seen as usable in the next function body.

umar456 commented:
I have updated the comments to indicate that the current version does not check the dominator frontier during validation. I was planning on adding this functionality later when validating the CFG. This PR is not intended to be a complete implementation of SSA validation.

dneto0 commented on source/validate.cpp:
```diff
@@ -261,6 +273,192 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+      break;
+
+    case SpvOpEnqueueKernel:
+      out = [](unsigned index) { return index == 7; };
```
The Invoke parameter is index==8

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+  ValidateInstructions(str, SPV_SUCCESS);
+}
+
+TEST_F(Validate, IdUndefinedBad) {
```
Bonus:  A module with no instructions at all is actually valid.
```
ValidateInstructions("", SPV_SUCCESS);
```

dneto0 replied:
My bad.  
Turns out OpMemoryModel is always required.   :-(

dneto0 commented on test/ValidateFixtures.h:
```diff
@@ -0,0 +1,52 @@
+  // Performs validation on the SPIR-V code and compares the result of the
+  // spvValidate function
+  void ValidateInstructions(std::string code, spv_result_t result);
+
```
For the other tests fixture, it was useful to capture the diagnostic so you could also check it.
(And remember to free it if it was allocated.  spvDiagnosticDestroy accepts a nullptr, so it's nice that way.)

Can be in the next CL of course.

dneto0 commented on test/ValidateFixtures.h:
```diff
@@ -0,0 +1,52 @@
+
+  // Performs validation on the SPIR-V code and compares the result of the
+  // spvValidate function
+  void ValidateInstructions(std::string code, spv_result_t result);
```
I think our preference has been to have the fixture methods return raw results, and then do the EXPECT_* calls in the main body of the text.  I believe that makes it easier to track down failures when they occur.

I'd like a 3rd opinion on this.

dmircevski replied:
+1

@dneto0 is exactly right.

When it's not possible for `EXPECT_*` to happen in the caller (eg, the callee returns `void`), then the callee's arguments should be `<<`ed into the `EXPECT` result, for better diagnostics -- see, for instance, `TextToBinaryTestBase::CompileSuccessfully()`.  Callers could always invoke `SCOPED_TRACE`, but that's clearly suboptimal.

umar456 replied:
Makes sense.

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+     OpExecutionMode %3 LocalSize 1 1 1
+%2 = OpTypeFunction %1
+%3 = OpFunction %1 None %2
+%1 = OpTypeVoid
```
This can also fail because you've put the void type decl after the start of the function body. So it depends on the order of check done inside the validator.  (Also, you could simplify by avoiding the entry point and function body.)
It would be better if you had just
```
OpMemoryModel Logical GLSL450
%1 = OpTypeFunction %2
%2 = OpTypeVoid
```

dneto0 replied:
Similarly, it's likely many later ID validation tests could be simplified by not having function declarations and bodies.
(Of course, for the CFG based ones you would need function bodies.)

dneto0 replied:
When the test code runs to several lines, I suggest adding an assembly comment to indicate which is the bad Id.

e.g.

```
%1 = OpTypeFunction %2       ; uses %2 before its definition
%2 = OpTypeVoid
```

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+  ValidateInstructions(str, SPV_SUCCESS);
+}
+
+TEST_F(Validate, ForwardNameMissingLabelBad) {
```
Nice test.

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+
+TEST_F(Validate, ForwardMemberNameGood) {
+  char str[] = R"(
+          OpMemberName %struct 0 "value"
```
Too bad, but needs an OpMemoryModel.   The principle is we want any test case to fail validation for exactly the effect under test.

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+  ValidateInstructions(str, SPV_ERROR_INVALID_ID);
+}
+
+TEST_F(Validate, ForwardGroupDecorateDecorationGroupDominateBad) {
```
Also need test case:

  OpGroupDecorate %missing %mat33 %mat44

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+  ValidateInstructions(str, SPV_ERROR_INVALID_ID);
+}
+
+TEST_F(Validate, ForwardFunctionCallGood) {
```
optional: could simplify this by using void(void) functions.
Don't need the int types or int constant values.

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+
+TEST_F(Validate, DISABLED_ForwardBranchConditionalGood) {
+  char str[] = R"(
+%voidt  = OpTypeVoid
```
OpMemoryModel yadda yadda yadda

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+%main   = OpFunction %voidt None %vfunct
+%true   = OpConstantTrue %boolt
+          OpSelectionMerge %endl None
+          OpBranchConditional %true %truel %falsel !1 !9
```
Aside: I think the assembler will accept branch weight literals that are unsigned 32-bit ints (without needing the ! syntax).  If you find different, please file an issue to me.

umar456 replied:
Yeah, it works without the ! syntax

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+%boolt  = OpTypeBool
+%vfunct = OpTypeFunction %voidt
+%main   = OpFunction %voidt None %vfunct
+%true   = OpConstantTrue %boolt
```
Constant definitions must occur before function bodies.  Constant definitions are in the same group as type definitions.

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+  ValidateInstructions(str, SPV_SUCCESS);
+}
+
+TEST_F(Validate, DISABLED_ForwardBranchConditionalWithWeightsGood) {
```
I think this test can be enabled.  The validator in its current state should let it pass.

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+%falsel = OpLabel
+          OpNop
+%endl   = OpLabel
+%true   = OpConstantTrue %boolt
```
This will fail in the future because the constants should all be defined before any function bodies.
Better to change this line to 
```
%true_copy = OpCopyObject %true
```
and change the previous OpBranchConditional to use %true_copy

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+  ValidateInstructions(str, SPV_ERROR_INVALID_ID);
+}
+
+TEST_F(Validate, ForwardBranchConditionalMissingLabelBad) {
```
(missing Target, yadda yadda yadda)

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+  ValidateInstructions(str, SPV_ERROR_INVALID_ID);
+}
+
+const string kBasicTypes = R"(
```
Nice

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+const string kBasicTypes = R"(
+%voidt  = OpTypeVoid
+%boolt  = OpTypeBool
+%int8t  = OpTypeInt 8 0
```
Universal validation rule says "Scalar integer types can be parameterized only as 32 bit, plus any additional sizes enabled by capabilities.
So any time you include "OpTypeInt 8  <anything>" you also need 
```OpCapability Int8``` right at the start of the module.

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+%vfunct = OpTypeFunction %voidt
+)";
+
+const string kKernelTypes = R"(
```
kKernelTypesAndConstants

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+)";
+
+const string kKernelDefinition = R"(
+%kfunct = OpTypeFunction %voidt %intptrt
```
Should move the function type definition into kKernelTypesAndConstants.
That's because later you might want to intermix yet more types before the function bodies start.

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+        OpFunctionEnd
+)";
+
+TEST_F(Validate, DISABLED_EnqueueKernelGood) {
```
I'm not clear on why this test has to be disabled in this MR.  I think it passes your current SSA checks. ? I haven't looked too closely.

umar456 replied:
This was failing because OpGetDefaultQueue is not correctly handled by ID validation. I disabled it to avoid committing failing tests. I have added functionality to bypass those validations for SSA tests.

Although now that I think about it that might not be a good idea. What is your opinion on it?

dneto0 replied:
Ok.  Just disable the test with a comment saying why.  That's fine.

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+  ValidateInstructions(str, SPV_SUCCESS);
+}
+
+TEST_F(Validate, DISABLED_ForwardEnqueueKernelGood) {
```
Similar.  I suspect this could be enabled.

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+
+TEST_F(Validate, ForwardEnqueueKernelNonDominantParameter1Bad) {
+  string str = forwardKernelNonDominantParameterBaseCode + R"(
+                    %err    = OpEnqueueKernel %uintt2 %dqueue %ndval %nevent %event %revent %kfunc %firstp %psize %palign %lsize
```
(Feel free to wrap the arguments on to subsequent lines.)

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+                    %err    = OpEnqueueKernel %uintt2 %dqueue %ndval %nevent %event %revent %kfunc %firstp %psize %palign %lsize
+                              OpReturn
+                              OpFunctionEnd
+                    %uintt2 = OpTypeInt 32 0
```
All type definitions precede function declarations and definitions.

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+%event  = OpConstantNull %eventt
+%revent = OpUndef %eventt
+
+%firstp = OpConstant %int8t 0
```
Constant definitions must go in the same group as type definitions.

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+
+INSTANTIATE_TEST_CASE_P(KernelArgs, Validate, ::testing::ValuesIn(cases));
+
+string SetOps(string &str, pair<string, bool> param) {
```
const string &str

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+
+string SetOps(string &str, pair<string, bool> param) {
+  regex placeholder("GET_KERNEL_OP");
+  string out = regex_replace(str, placeholder, param.first);
```
I think this will fail with GNU g++ 4.8.4 since the GNU libc++ regex library is broken.  :-(

dneto0 replied:
I think you can accomplish what you want with std::stringstream and its << operator.
And to avoid repeating yourself, you can put the conditional logic in a function that takes a std::ostream* parameter to modify.

umar456 replied:
Yeah, That was my first iteration but this was so much more elegant. :-) :-(

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+  return out2;
+}
+
+TEST_P(Validate, DISABLED_GetKernelGood) {
```
I think this can be enabled?

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+  ValidateInstructions(out, SPV_SUCCESS);
+}
+
+TEST_P(Validate, DISABLED_ForwardGetKernelGood) {
```
I think this can be enabled?

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+                  %numsg   = GET_KERNEL_OP %uintt2 %ndval %kfunc %firstp %psize %palign
+                             OpReturn
+                             OpFunctionEnd
+                  %uintt2  = OpTypeInt 32 0
```
types go before functions.

dneto0 commented:
Ok, finished review of latest commit.

* Must fix the index == 8 bug in checking for allowable forward use of Id with OpFunction
* Should fix tests so they only fail for the reason being tested. A lots of the test cases are invalid for multiple reasons, even though the validator will only check on of those reasons right now.  For example a lot of the tests are missing OpMemoryModel, or have OpConstant or OpTypeInt that occur after the first OpFunction instruction.
* Nice to fix:  Simplify tests to remove extraneous stuff.  E.g. a lot of Id checks can be done without any function bodies at all.  I've suggested one in detail.
* Quibble:  Some naming things.  Like instead of "...MissingLabel..." say "...MissingTarget..." since "Label" is already a first class concept in SPIR-V.

(and maybe other stuff I've forgotten by now.)

dneto0 commented:
Oh, and I forgot to thank you for making thorough checks for invalidity of each ID operand that can't be forward declared!

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,699 @@
+
+  if (!param.second) {
+    regex ndrange_regex("[^\n]%ndval");
+    out2 = regex_replace(out, ndrange_regex, "");
```
This doesn't compile with GCC g++ 4.8.4

Validate.SSA.cpp:592:48: error: no matching function for call to ‘regex_replace(std::string&, std::regex&, const char [1])’
and later details say template argument deduction/substitution failed.

umar456 replied:
What compilers are we targeting?

dneto0 replied:
VisualStudio 2013. (e.g. no constexpr)
GCC 4.8.2 ; that's what comes with Ubunto 14.04 Trusty LTS.
A recent Clang, but I expect that will be at least as feature rich as the GCC.

I should document this in the README.  :-(

FYI.  The Chromium project notes gotchas with various compiler's C++11 support:
https://chromium-cpp.appspot.com/
E.g. you can see there that constexpr isn't supported in VisualStudio 2013
But we use more features of C++11 than Chromium allows, I think.

umar456 replied:
Can I add a dependency to boost.regex or boost.xpressive?

umar456 commented:
Addressed the feedback thus far.

umar456 commented:
* Added additional unit tests (OpPhi, OpLoopMerge)
* Added support for OpName based diagnostics
* Unit tests now check for id names in diagnostic message

dneto0 commented:
Whoops, I was reviewing an older commit.   Still looking...

dneto0 commented:
Done this round.
Issues:
*  Fix the block termination in the OpPhi tests
* Any ID operand to OpLoopMerge can be forward declared.   Either add a TODO for tests for OpLoopMerge or add tests themselves.
* Missing blurbs in a couple of places.  (May as well do it in the PR.)

Recommend not adding new tests to this PR because they can potentially require further rounds of review.

umar456 commented:
Addressed feedback so far.

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,1000 @@
+    )";
+  CompileSuccessfully(str);
+  ASSERT_EQ(SPV_ERROR_INVALID_ID, ValidateInstructions());
+  ASSERT_TRUE(ContainsString(getDiagnosticString(), "missing"));
```
OPTIONAL:  Sorry I didn't think of this before.  But we tend to use gmock matchers, so this would be:

```
ASSERT_THAT(getDiagnosticString(), HasSubstr("missing"));
```

(assuming early in the file you have:
```
#include "gmock/gmock.h"
using ::testing::HasSubstr
```
)

See https://github.com/google/googletest/blob/master/googlemock/docs/CheatSheet.md

It can be very powerful since there are so many predicates to apply.  In particular, it's great at matching containers, with very good messages when things fail.

Good for future tests you write.

dneto0 commented on test/Validate.SSA.cpp:
```diff
@@ -0,0 +1,1000 @@
+
+const string kBasicTypes = R"(
+           OpCapability Int8
+           OpCapability DeviceEnqueue
```
Good catch!

dneto0 commented:
I'm satisfied with this for now.   I plan to rebase and squash before pushing onto master.

Also, I may have a few more OpPhi cases in mind.  But they'll be fore later CLs.

dneto0 commented:
+2

dneto0 commented:
Rebased, and made the following changes:
*  Avoid std::to_string because that's not available on Android.  Reworked some local code to use std::strinstream instead.
* Did "using std::placeholders::_1" instead of blanket "using namespace std::placeholders" as per Google C++ style guide
* Wrapped the local functions in anonymous namespace.  Avoids polluting the global namespace.
* For diagnostics generation, changed a C-style cast to uint64_t to static cast to size_t, to fix compilation on Android.
* Clang-format

dneto0 commented:
Squashed and pushed onto master.
